### PR TITLE
Add onboarding API endpoint

### PR DESCRIPTION
**Pull request by KrauseFx**
- [x] We need to protect the API endpoint

Fixes https://github.com/fastlane/ci/issues/1012

### CODE REVIEW

File: app/features-json/setup_json_controller.rb [hunk 1]
```diff
@@ -8,5 +8,109 @@ class SetupJSONController < APIController
...
       return json(Services.onboarding_service.correct_setup?)
     end
+
+    # TODO: we need to protect this API endpoint
+    # one the onboarding user flow, we don't have any accounts yet
+    # however, we need to prohibit sending a POST request to this API endpoint
+    # after CI was already set up
+    post HOME.to_s, authenticate: false do
+      # we set this as default so that we don't need duplicate code for error detection
+      params["bot_account"] ||= {}
```
Comment (nakhbari) [inline app/features-json/setup_json_controller.rb]: nit: maybe a new var so later on we don't have to do `params[..]`?

File: app/features-json/setup_json_controller.rb [hunk 2]
```diff
@@ -8,5 +8,109 @@ class SetupJSONController < APIController
...
+
+      encryption_key = params["encryption_key"]
+      bot_token = params["bot_account"]["token"]
+      bot_password = params["bot_account"]["password"] # TODO: What do we need this for
+      config_repo = params["config_repo"]
+      initial_onboarding_token = params["initial_onboarding_token"]
```
Comment (nakhbari) [inline app/features-json/setup_json_controller.rb]: I never understood what the initial onboarding token was. Why don't we just use the bot token?

File: app/features-json/setup_json_controller.rb [hunk 3]
```diff
@@ -8,5 +8,109 @@ class SetupJSONController < APIController
...
+      validate_api_token_correct!(initial_onboarding_token)
+      dot_keys_values[:initial_onboarding_user_api_token] = initial_onboarding_token
+
+      # Step: Validate encryption key
+      validate_encryption_key!(encryption_key)
+      dot_keys_values[:encryption_key] = encryption_key
+
+      # Step: Validate repo URL
+      # TODO: What do we want to validate here?
+      dot_keys_values[:repo_url] = config_repo
```
Comment (nakhbari) [inline app/features-json/setup_json_controller.rb]: should we validate that the bot has access to the repo?

File: app/features-json/setup_json_controller.rb [hunk 4]
```diff
@@ -8,5 +8,109 @@ class SetupJSONController < APIController
...
+          error_message: error_message,
+          error_key: "Onboarding.Token.MissingScope"
+        )
+      end
+    end
+
+    def validate_encryption_key!(encryption_key)
+      if encryption_key.to_s.length < 5
+        json_error!(
+          error_message: "Encryption key not long enough, it must be at least 5 characters",
```
Comment (nakhbari) [inline app/features-json/setup_json_controller.rb]: oh i didn't know this was a requirement, we should add it to the UI to make sure it's at least 5 chars

File: app/features-json/setup_json_controller.rb (hunk 2 above)
Comment (KrauseFx) [reply app/features-json/setup_json_controller.rb]: I think you're right, we could do that 👍 Will update the existing code base also to reflect this. Anything we're missing @taquitos

File: app/features-json/setup_json_controller.rb (hunk 1 above)
Comment (KrauseFx) [reply app/features-json/setup_json_controller.rb]: I only default to `{}` to easily access nested hashes later in the code. We can also update the code base to use `.fetch`

File: app/features-json/setup_json_controller.rb [hunk 5]
```diff
@@ -8,5 +8,150 @@ class SetupJSONController < APIController
...
+      bot_password = params["bot_account"]["password"] # TODO: What do we need this for
+      config_repo = params["config_repo"].to_s
+      initial_onboarding_token = params["initial_onboarding_token"]
+
+      missing_parameters = {
+        encryption_key: encryption_key,
+        bot_token: bot_token,
+        bot_password: bot_password,
+        config_repo: config_repo,
+        initial_onboarding_token: initial_onboarding_token
```
Comment (nakhbari) [inline app/features-json/setup_json_controller.rb]: so can we remove the requirement of this param?

>nakhbari:
I never understood what the initial onboarding token was. Why don't we just use the bot token?
>KrauseFx:
I think you're right, we could do that 👍 Will update the existing code base also to reflect this. Anything we're missing @taquitos

Comment (KrauseFx) [reply app/features-json/setup_json_controller.rb]: So seems like for now we decided to leave it as is, and do the refactor in a separate PR, right?

File: app/features-json/setup_json_controller.rb [hunk 6]
```diff
@@ -8,5 +8,150 @@ class SetupJSONController < APIController
...
+    # one the onboarding user flow, we don't have any accounts yet
+    # however, we need to prohibit sending a POST request to this API endpoint
+    # after CI was already set up
+    post HOME.to_s, authenticate: false do
+      # we set this as default so that we don't need duplicate code for error detection
+      params["bot_account"] ||= {}
+
+      encryption_key = params["encryption_key"]
+      bot_token = params["bot_account"]["token"]
+      bot_password = params["bot_account"]["password"] # TODO: What do we need this for
```
Comment (KrauseFx) [inline app/features-json/setup_json_controller.rb]: @nakhbari what is this value being used for?

**Comment by nakhbari** [comment]
I marked this as blocked since we're ironing out the details of Onboarding with the new Github OAuth App

**Comment by KrauseFx** [comment]
I think it's safe to say that this change will not get merged, as we have a new approach for onboarding + auth. I'll close this PR for now 👍

**Pull request closed without merging**